File: Recursion Maze/Sliding Window/Q1. Contains Duplicate II/Q1_ContainsDuplicateII.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        // Map to store the last index of each number
        Map<Integer, Integer> map = new HashMap<>();
        
        // Loop through array
        for (int i = 0; i < nums.length; i++) {
            int num = nums[i];
            
            // If num was seen before and difference of indices <= k
            if (map.containsKey(num) && i - map.get(num) <= k) {
                return true;
            }
            
            // Update the index of the current number
            map.put(num, i);
        }
        
        // No valid pair found
        return false;
    }
}

File: Recursion Maze/Sliding Window/Q1. Contains Duplicate II/Q1_ContainsDuplicateII.py
class Solution(object):
    def containsNearbyDuplicate(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: bool
        """
        seen = {}  # Dictionary to store number -> last index

        for i, num in enumerate(nums):
            # If number was seen before and distance between indices <= k
            if num in seen and i - seen[num] <= k:
                return True

            # Update last index of the number
            seen[num] = i

        # No such pair found
        return False
